packages/flutter_tools/lib/src/version.dart

```diff
@@ -72,11 +81,59 @@ class FlutterVersion {
   }
 
   /// A date String describing the last framework commit.
-  static String get frameworkCommitDate {
-    return _runSync(<String>['git', 'log', '-n', '1', '--pretty=format:%ad', '--date=format:%Y-%m-%d %H:%M:%S'], Cache.flutterRoot);
+  static String get frameworkCommitDate => _latestGitCommitDate();
+
+  static String _latestGitCommitDate([String branch]) {
+    final List<String> args = <String>['git', 'log'];
+
+    if (branch != null) {
+      args.add(branch);
+    }
+
+    args.addAll(<String>['-n', '1', '--pretty=format:%ad', '--date=format:%Y-%m-%d %H:%M:%S']);
+    return _runSync(args, Cache.flutterRoot, lenient: false);
   }
 
-  static FlutterVersion get instance => context.putIfAbsent(FlutterVersion, () => new FlutterVersion._());
+  /// The name of the temporary git remote used to check for the latest
+  /// available Flutter framework version.
+  ///
+  /// In the absence of bugs and crashes a Flutter developer should never see
+  /// this remote appear in their `git remote` list, but also if it happens to
+  /// persist we do the proper clean-up for extra robustness.
+  static const String _kVersionCheckRemote = '__flutter_version_check__';
+
+  /// The date of the latest framework commit in the remote repository.
+  ///
+  /// Throws [ToolExit] if a git command fails, for example, when the remote git
+  /// repository is not reachable due to a network issue.
+  static String get remoteFrameworkCommitDate {
```

Comment:
We'd like to get away from using `runSync()` (#7102), so for new API, we should definitely make it return a Future and using `run()`